sys/net/gnrc/link_layer/netdev2/gnrc_netdev2_eth.c

```diff
@@ -0,0 +1,195 @@
+#include "net/gnrc.h"
+#include "net/gnrc/gnrc_netdev2.h"
+#include "net/ethernet/hdr.h"
+
+#include "od.h"
+
+#define ENABLE_DEBUG (0)
+#include "debug.h"
+
+static gnrc_pktsnip_t *_recv(gnrc_netdev2_t *gnrc_netdev2)
+{
+    netdev2_t *dev = gnrc_netdev2->dev;
+    int bytes_expected = dev->driver->recv(dev, NULL, 0);
+    gnrc_pktsnip_t *pkt = NULL;
+    gnrc_pktsnip_t *whole_pkt = NULL;
+
+    if (bytes_expected) {
+        pkt = gnrc_pktbuf_add(NULL, NULL,
+                bytes_expected,
+                GNRC_NETTYPE_UNDEF);
+
+        if(!pkt) {
+            DEBUG("_recv_ethernet_packet: cannot allocate pktsnip.\n");
+            goto out;
+        }
+
+        whole_pkt = pkt;
+
+        int nread = dev->driver->recv(dev, pkt->data, bytes_expected);
+        if(nread <= 0) {
+            DEBUG("_recv_ethernet_packet: read error.\n");
+            goto safe_out;
+        }
+
+        size_t data_len = (nread - sizeof(ethernet_hdr_t));
+
+        if (nread < bytes_expected) {
+            /* we've got less then the expected packet size,
+             * so free the unused space.*/
+
+            DEBUG("_recv_ethernet_packet: reallocating.\n");
+            gnrc_pktbuf_realloc_data(pkt, nread);
+        }
+
+        ethernet_hdr_t *hdr = (ethernet_hdr_t *)pkt->data;
+
+        gnrc_nettype_t receive_type =
+            gnrc_nettype_from_ethertype(byteorder_ntohs(hdr->type));
+
+        /* create netif header */
+        gnrc_pktsnip_t *netif_hdr;
+        netif_hdr = gnrc_pktbuf_add(NULL, NULL,
+                sizeof(gnrc_netif_hdr_t) + (2 * ETHERNET_ADDR_LEN),
+                GNRC_NETTYPE_NETIF);
+
+        if (netif_hdr == NULL) {
+            DEBUG("gnrc_netdev2_eth: no space left in packet buffer\n");
+            goto safe_out;
+        }
+
+        gnrc_netif_hdr_init(netif_hdr->data, ETHERNET_ADDR_LEN, ETHERNET_ADDR_LEN);
+        gnrc_netif_hdr_set_src_addr(netif_hdr->data, hdr->src, ETHERNET_ADDR_LEN);
+        gnrc_netif_hdr_set_dst_addr(netif_hdr->data, hdr->dst, ETHERNET_ADDR_LEN);
+        ((gnrc_netif_hdr_t *)netif_hdr->data)->if_pid = thread_getpid();
+
+        DEBUG("gnrc_netdev2_eth: received packet from %02x:%02x:%02x:%02x:%02x:%02x "
+                "of length %zu\n",
+                hdr->src[0], hdr->src[1], hdr->src[2], hdr->src[3], hdr->src[4],
+                hdr->src[5], nread);
+#if defined(MODULE_OD) && ENABLE_DEBUG
+        od_hex_dump(hdr, nread, OD_WIDTH_DEFAULT);
+#endif
+
+        /* Mark netif header and payload for next layer */
+        if ((pkt = gnrc_pktbuf_add(netif_hdr, ((char*)pkt->data) + sizeof(ethernet_hdr_t),
+                                   data_len, receive_type)) == NULL) {
+            DEBUG("gnrc_netdev2_eth: no space left in packet buffer\n");
+            gnrc_pktbuf_release(netif_hdr);
+        }
+
+        goto out;
+    }
+
+safe_out:
+    gnrc_pktbuf_release(pkt);
```

Comment:
This potentially releases NULL pointers which does not fulfill the function's precondition.